egs/wsj/s5/steps/cleanup/internal/align_ctm_ref.py

```diff
@@ -0,0 +1,213 @@
+import sys, argparse, imp
+
+verbosity = 0
+
+def GetArgs():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--hyp-format", type=str, choices = ["Text", "CTM"],
+                        "Format used for the hypothesis")
+    parser.add_argument("--eps-symbol", type=str, default = "-",
+                        "Symbol used to contain alignment to empty symbol")
+    parser.add_argument("--verbose", type=int, default=0)
+    parser.add_argument("--correct-score", type=int, default=1,
+                        "Score for correct matches")
+    parser.add_argument("--substitution-penalty", type=int, default=1,
+                        "Penalty for substitution errors")
+    parser.add_argument("--deletion-penalty-type", type=str, default = "linear",
+                        choices = ["linear"],
+                        help = "Type of function for deletion penalty")
+    parser.add_argument("--deletion-penalty-scale", type-int, default=1,
+                        "Scale on deletion penalty")
+    parser.add_argument("--insertion-penalty-type", type=str, default = "linear",
+                        choices = ["linear"],
+                        help = "Type of function for insertion penalty")
+    parser.add_argument("--insertion-penalty-scale", type-int, default=1,
+                        "Scale on insertion penalty")
+    parser.add_argument("--debug-only", type=str, default = "false",
+                        choices = ["true", "false"],
+                        help = "Run test functions only")
+    parser.add_argument("ref", type=str,
+                        help = "Reference text")
+    parser.add_argument("hyp", type=str,
+                        help = "Hypothesis text or CTM")
+
+    args = parser.parse_args()
+    verbosity = args.verbose
+    args.debug_only = True if args.debug_only == "true" else False
+
+    return args
+
+def SmithWatermanAlignment(ref, hyp, eps_symbol,
+                           SimilarityScore, DelScore, InsScore,
+                           hyp_format = "CTM"):
+    output = []
+
+    M = len(ref)
+    N = len(hyp)
+
+    bp = [ [ ] for x in range(M+1) ]
+    H = [ [ ] for x in range(M+1) ]
+
+    for m in range(M+1):
+        H[m] = [ 0 for x in range(N+1) ]
+        bp[m] = [ (0,0) for x in range(N+1) ]
+
+    global_max = 0
+    global_max_element = (0,0)
+
+    for m in range(1, M+1):
+        for n in range(1, N+1):
+            sub_or_ok = (H[m-1][n-1] +
+                         SimilarityScore(ref[m-1],
+                                         hyp[n-1][3] if hyp_format == "CTM"
+                                                   else hyp[n-1]))
+            if sub_or_ok > 0:
+                H[m][n] = sub_or_ok
+                bp[m][n] = (m-1, n-1)
+                if verbosity > 2:
+                    print ("({0},{1}) -> ({2},{3}): {4} ({5},{6})".format(m-1,n-1,m,n, H[m][n], ref[m-1], hyp[n-1]), file = sys.stderr)
+
+            for k in range(1, m+1):
```

Comment:
It occurred to me that the only reason this algorithm needs to be cubic (instead of quadratic) in the length of the sequence is that it supports non-linear functions for the deletion and insertion cost.  Since we're probably not going to use these anyway, it would be much better to remove that capability and use a quadratic algorithm.  Really the algorithm can be the same as the one used for Levenshtein edit distance, except with variable ins/del/sub costs, and instead of initializing position (0,0) as zero cost [and the rest infinity], and then evaluating the best cost at the end (N, M), you initialize all positions (n,0) to zero and then evaluate the best cost at all positions (n, M).